### PR TITLE
Expand `BoundedContextBuilder` API

### PR DESCRIPTION
**Pull request by dmitrykuzmin**
This PR addresses #1099.

The `BoundedContextBuilder` has been expanded with the API allowing to:
1. Register/unregister repositories by the entity class.
2. Register/unregister `CommandDispatcher`s.
3. Register/unregister `EventDispatcher`s.

The dispatcher registration works the same as repository registration, i.e. dispatchers are added to the `BoundedContextBuilder` but registered within `BoundedContext` only after the `build()` method is called.

If a `Repository` is passed in place of command or event dispatcher, it is still registered as a repository. The same now applies to the `BlackBoxBoundedContext` methods.

### CODE REVIEW

**Comment by dmitrykuzmin** [comment]
@armiol PTAL.

File: server/src/main/java/io/spine/server/event/EventBus.java [hunk 1]
```diff
@@ -144,6 +144,15 @@ boolean hasDispatchers(EventClass eventClass) {
         return !dispatchers.isEmpty();
     }
 
+    /**
+     * Obtains a view {@code Set} of events that are known to this {@code EventBus}.
+     *
+     * <p>This set is changed when event dispatchers or handlers are registered or un-registered.
+     */
+    public Set<EventClass> registeredEventClasses() {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/event/EventBus.java]: Is this a part of an internal or public API? Looks very internal to me, similar to `hasDispatchers(..)` above.

File: server/src/main/java/io/spine/server/event/EventDispatcherRegistry.java [hunk 2]
```diff
@@ -66,6 +66,17 @@ public void unregister(EventDispatcher<?> dispatcher) {
         return super.getDispatchersForType(messageClass);
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Overrides to expose the method to
+     * {@link EventBus#registeredEventClasses() EventBus}.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/event/EventDispatcherRegistry.java]: The contents of this line fit the previous one.

File: server/src/test/java/io/spine/server/bc/given/Given.java [hunk 3]
```diff
@@ -53,4 +66,56 @@ public static BcProjectStarted projectStarted(ProjectId id) {
                                    .build();
         }
     }
+
+    public static class NoopCommandDispatcher implements CommandDispatcher<ProjectId> {
```
Comment (armiol) [inline server/src/test/java/io/spine/server/bc/given/Given.java]: We usually use `NoOp...` spelling, not `Noop`. I can see just a single `Noop` usage in the project (see `NoopQueryProcessor`), which also needs to be addressed (maybe you could help?)

File: server/src/main/java/io/spine/server/BoundedContextBuilder.java [hunk 4]
```diff
@@ -302,6 +338,84 @@ boolean hasRepository(Repository<?, ?> repository) {
...
+                                                              .equals(entityClass));
+        return result;
+    }
+
+    /**
+     * Adds the passed command dispatcher to the dispatcher registration list which will be
+     * processed after the Bounded Context is created.
+     */
+    @CanIgnoreReturnValue
+    public BoundedContextBuilder addCommandDispatcher(CommandDispatcher<?> commandDispatcher) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/BoundedContextBuilder.java]: The tricky business here is that repositories are also `CommandDispatcher`s and `EventDispatcher`s. 

Let's make it impossible to use `[add | remove | has ]...Dispatcher(...)` and ` API with the `...Repository`-typed arguments.

**Comment by dmitrykuzmin** [comment]
@armiol PTAL again.

**Comment by dmitrykuzmin** [comment]
@armiol PTAL again.

**Pull request merged**